### PR TITLE
Add tag support to the consultant builder

### PR DESCRIPTION
**Pull request by rogierslag**
This add service instance tag support to Consultant. It allows us for example to tag service instances with git sha's, whether production can be routed and any other relevant meta-data.

Also did an autoformat which decided that some comments really needed new alignment.

@michaeldejong Can you check whether you agree with these changes?

### CODE REVIEW

File: src/main/java/me/magnet/consultant/Consultant.java [hunk 1]
```diff
@@ -395,8 +413,11 @@ public void registerService(int port) {
 			String serviceId = id.getInstance().get();
 			String serviceName = id.getServiceName();
 			String serviceHost = id.getHostName().orElse(InetAddress.getLoopbackAddress().getHostAddress());
-			Check check = new Check("http://" + serviceHost + ":" + port + "/" + healthEndpoint, HEALTH_CHECK_INTERVAL);
-			ServiceRegistration registration = new ServiceRegistration(serviceId, serviceName, serviceHost, port, check);
+			Collection<String> tags = id.getTags();
+			Check check = new Check("http://" + serviceHost + ":" + port + "/" + healthEndpoint,
+					HEALTH_CHECK_INTERVAL);
+			ServiceRegistration registration = new ServiceRegistration(serviceId, serviceName, serviceHost, port,
+					check, tags.toArray(new String[0]));
```
Comment (AlexKolpa) [inline src/main/java/me/magnet/consultant/Consultant.java]: The `new String[0]` just leads to a useless allocation. Do `new String[tags.size()]` instead.

**Comment by AlexKolpa** [comment]
>Also did an autoformat which decided that some comments really needed new alignment.

As is tradition

Besides the String array comment, code looks good :+1:

**Comment by rogierslag** [comment]
@michaeldejong Tests fail, but those do not trigger a test failure. Can you have a quick look?

**Comment by AlexKolpa** [comment]
Looks like the mocked entity that's being returned in the test should be wrapped as an array, instead of just the key-value object.

**Comment by alexnederlof** [comment]
:+1: but revert all the format changes

File: src/main/java/me/magnet/consultant/Consultant.java [hunk 2]
```diff
@@ -64,6 +65,7 @@
 		private final SetMultimap<String, SettingListener> settingListeners;
 		private final Set<ConfigListener> configListeners;
 		private String healthEndpoint;
+		private Collection<String> tags;
```
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/Consultant.java]: This should be a `Set`.

File: src/main/java/me/magnet/consultant/Consultant.java [hunk 3]
```diff
@@ -167,6 +170,20 @@ public Builder identifyAs(String serviceName, String datacenter, String hostname
...
 		/**
+		 * Add a service tag to be registered with Consul. This can be used to specify any information relevant to the
+		 * specific instance. Examples can be the version of a service or whether a service is ready for production
+		 * traffic
+		 *
+		 * @param tag The value of the tag
+		 * @return The Builder instance
+		 */
+		public Builder withTag(String tag) {
+			checkArgument(!this.tags.contains(tag), "This tag has already been specified");
```
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/Consultant.java]: Not needed, if you use a `Set`.

File: src/main/java/me/magnet/consultant/ServiceIdentifier.java [hunk 4]
```diff
@@ -47,6 +49,12 @@ public String getServiceName() {
 		return instance;
 	}
 
+	public Collection<String> getTags() { return tags; }
+
+	public void addTagSet(Collection<String> newTags) {
```
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/ServiceIdentifier.java]: This object is meant to be immutable. So no setter methods allowed.

File: src/main/java/me/magnet/consultant/ServiceIdentifier.java [hunk 5]
```diff
@@ -47,6 +49,12 @@ public String getServiceName() {
 		return instance;
 	}
 
+	public Collection<String> getTags() { return tags; }
```
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/ServiceIdentifier.java]: Return a `Set`.

File: src/main/java/me/magnet/consultant/ServiceIdentifier.java [hunk 6]
```diff
@@ -114,12 +123,13 @@ public int hashCode() {
...
 				.append(hostName)
 				.append(instance)
+				.append(tags)
 				.toHashCode();
 	}
 
 	@Override
 	public String toString() {
-		List<String> descriptors = Lists.newArrayList();
+		Collection<String> descriptors = Lists.newArrayList();
```
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/ServiceIdentifier.java]: Why this?

**Comment by michaeldejong** [comment]
:+1: if you fix comments, and revert formatting changes to docs.

Btw, what is the reason we need this? Or is it just added for the lulz?

**Comment by AlexKolpa** [comment]
@michaeldejong Since consul supports tags, we can have some running service version information there (like the git head commit). A few days ago we realized we had been running an older version of news, which was causing errors (missing enum values), which was the reason @rogierslag deciced to set this up.

**Comment by michaeldejong** [comment]
That's fine, but it would have been nice if this use-case was stated in the PR OP. It's a public repo after all.

**Comment by michaeldejong** [comment]
@rogierslag can you patch this PR?

File: src/main/java/me/magnet/consultant/Consultant.java [hunk 7]
```diff
@@ -395,8 +413,11 @@ public void registerService(int port) {
 			String serviceId = id.getInstance().get();
 			String serviceName = id.getServiceName();
 			String serviceHost = id.getHostName().orElse(InetAddress.getLoopbackAddress().getHostAddress());
-			Check check = new Check("http://" + serviceHost + ":" + port + "/" + healthEndpoint, HEALTH_CHECK_INTERVAL);
-			ServiceRegistration registration = new ServiceRegistration(serviceId, serviceName, serviceHost, port, check);
+			Collection<String> tags = id.getTags();
+			Check check = new Check("http://" + serviceHost + ":" + port + "/" + healthEndpoint,
+					HEALTH_CHECK_INTERVAL);
+			ServiceRegistration registration = new ServiceRegistration(serviceId, serviceName, serviceHost, port,
+					check, tags.toArray(new String[tags.size()]));
```
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/Consultant.java]: Just make it a `Set` in the `ServiceRegistration`.

**Comment by michaeldejong** [comment]
Also can you make sure we can set (additional) tags with the environment variable `SERVICE_TAGS` (comma-separated).

**Comment by rogierslag** [comment]
Yup will do a fixup/rewrite of this logic tomorrow

**Pull request closed without merging**